stablehlo/transforms/StablehloCanonicalizeDynamism.cpp

```diff
@@ -34,6 +35,54 @@ namespace stablehlo {
 
 namespace {
 
+struct CanonicalizeCustomCallOpPattern : public OpRewritePattern<CustomCallOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(CustomCallOp op,
+                                PatternRewriter& rewriter) const override {
+    SmallVector<ShapedTypeComponents> refinements;
+    if (failed(hlo::getShapeRefinements(op.getLoc(), op, refinements)))
+      return rewriter.notifyMatchFailure(op, "expected valid refinements");
+    auto indicesAttr =
+        op->getAttr("indices_of_shape_operands").cast<DenseIntElementsAttr>();
+    DenseSet<int64_t> indices(indicesAttr.value_begin<int64_t>(),
+                              indicesAttr.value_end<int64_t>());
+
+    // Discard the indices_of_shape_operands attribute.
+    // We rely on the verification logic implemented in getShapeRefinements to
+    // make sure that its value are consistent with the result type.
+    // In the future, when we upgrade indices_of_shape_operands from an
+    // experiment to a full-fledged StableHLO feature, this logic will be moved
+    // to a proper verifier.
+    SmallVector<NamedAttribute> newAttrs;
+    for (auto attr : op->getAttrs()) {
+      if (attr.getName() == "indices_of_shape_operands") continue;
+      newAttrs.push_back(attr);
+    }
+
+    // Discard the operands that correspond to indices_of_shape_operands.
+    // Similarly, we rely on the verification logic implemented in
+    // getShapeRefinements to make sure that indices_of_shape_operands is
+    // consistent with these operands.
+    SmallVector<Value> newOperands;
+    auto resultIndex = 0;
+    for (auto& operand : op->getOpOperands()) {
```

Comment:
`llvm::enumerate`?

Reply:
getOpOperands already includes an equivalent of llvm::enumerate, since its results have `getOperandNumber`. If you meant to ask whether we can avoid having the `resultIndex` variable, I don't think we can because this variable is not incremented on every loop iteration.